### ISSUE TITLE
do_task should return the task's return value

### ISSUE THREAD

**Issue by mbroadhead** [RexOps/Rex#706]
I would expect do_task() to return the task's return value, just like run_task() does.

$ rex -v
(R)?ex 1.2.1

$ uname -a
Darwin user.local 14.3.0 Darwin Kernel Version 14.3.0: Mon Mar 23 11:59:05 PDT 2015; root:xnu-2782.20.48~5/RELEASE_X86_64 x86_64

Consider the following test:

```perl
# file: return_values.t
use Rex -base;
use Test::More tests => 3;

my $got_return_value;
my $expected_return_value = { foo => "bar" };

task "foo" => sub {
  return { foo => "bar" };
};

$got_return_value = run_task "foo";
is_deeply( $got_return_value, $expected_return_value, "Calling task with run_task() returns the task's return value" );

$got_return_value = do_task "foo";
is_deeply( $got_return_value, $expected_return_value, "Calling task with do_task() returns the task's return value" );

$got_return_value = foo();
is_deeply( $got_return_value, $expected_return_value, "Calling task by function returns the task's return value" );
```

```
$ perl return_values.t
1..3
ok 1 - Calling task with run_task() returns the task's return value
[2015-05-26 11:56:20] INFO - Running task foo on <local>
not ok 2 - Calling task with do_task() returns the task's return value
#   Failed test 'Calling task with do_task() returns the task's return value'
#   at return_values.t line 15.
#     Structures begin differing at:
#          $got = undef
#     $expected = HASH(0x7fc088804638)
[2015-05-26 11:56:20] INFO - Running task foo on current connection
ok 3 - Calling task by function returns the task's return value
# Looks like you failed 1 test of 3.
```

**Comment by ferki** [comment]
@mbroadhead: good point! Are you considering to send a pull request about the return values? You might also be interested in #697 and #691, as those issues are also dealing (partly) with return values.

**Comment by mbroadhead** [comment]
Sure I'll look into this soon and get you a pull request. I just figured
I'd get an issue posted while I'm thinking about it.

**Comment by krimdomu** [comment]
@mbroadhead @ferki but what should do_task return?
do_task executes the task on all given hosts (for example on a complete group) how should the return value look like?

**Comment by mbroadhead** [comment]
@krimdomu You have a good point. do_task can also run multiple tasks at once. In this case, what would be the expected return value?
I was looking into this last night and it seems do_task() calls Rex::TaskList->run($task) for each task passed in. This method calls the distribution class's run method, presumably Rex::TaskList::Base->run(), which loops through each server, forks, and finally calls Rex::Task->run().

I'm now not sure it makes sense to return the task's return value since there might be multiple servers and/or tasks being run within one invocation of do_task. What are your thoughts @ferki @krimdomu ?

**Comment by krimdomu** [comment]
one option i had thought about it in the past already is to return a hash which contains all the return values for all servers. For example:

```perl
my $ret = do_task[qw/task1 task2/];
$ret = {
  "task1" => {
    "server1" => "foo",
    "server2" => "bar",
  },
  "task2" => {
    "server1" => 5,
    "server2" => 5,
  },
};
```

**Comment by mbroadhead** [comment]
@krimdomu I think your idea of returning a hash of `{ "task1" => { "server1" => "foo", .. }, .. }` is the best idea. Do you guys agree @ferki @krimdomu ? If so, I will start looking into implementing this.

**Comment by ferki** [comment]
@ferki: well, I can't come up with a better idea, so I'd vote for let's see how it works in practice.
@krimdomu: we might need to control this behaviour with a feature flag, right?

**Comment by kablamo** [comment]
Correct me if I'm wrong, but I think a hash doesn't work in the edge case where you have
```
do_task [qw/task1 task2 task1/];
```
For that situation you would need an arrayref.

**Comment by kablamo** [comment]
This is way more complicated than it looks on the surface.  There are at least 4 separate questions present in this issue.  I will try to answer each one.

**1. Why are there two methods for running tasks?***  Correct me if I'm wrong -- I believe the fundamental problem here and the reason this is so confusing to users is that there are two method for running tasks.  There should just be one.   I get that `do_task()` and `run_task()` have different features and do different things, but their feature sets should be merged and one of the methods should be deprecated. 

**2. How should users return data from a task?**
As @ferki and I discussed a few days ago, its complicated for `do_task()` to return the task value because unlike `run_task()` it has the ability to fork and run tasks in parallel.  This means the return value has to get passed from the child process to the parent via some sort of IPC mechanism.  Which won't work for complex objects.  But we can make it work for simple data structures by doing something tricky like this: https://metacpan.org/pod/Parallel::ForkManager#RETRIEVING-DATASTRUCTURES-from-child-processes 

**3. How should we indicate if a task succeeded or failed?**
*Option 1:* Return true/false from every task to indicate success/failure.  This is fine.  But it means users can't return data from their task.  Instead they would need to use `Rex::Shared::Vars` which is also fine but I think its not as intuitive as just returning your data from the task.  It also means users need to change their tasks so they explicitly return true/false.  

*Option 2:*  We define a successful task as one that doesn't die.  This will work especially well with people who use `exec_autodie`.  It also means users can return data from their task.  This is the solution I prefer. 

**4. What should `run_task()` and `do_task()` return?**
Here I assume you agree with my answer to question 3 above.

In the case of `do_task()`, users might be interested in at least 3 kinds of data: 
 1. The data returned from the task.  
 2. Whether the task failed or succeeded.  
 3. Which server the task ran on -- remember that `do_task()` forks a child process for each server it runs against.

*Option 1:* Return the data from the task.  Die if the task failed on any of the servers.  If you run the task against 10 servers, and the first 9 succeeded and the last one dies, the user would lose all the data returned from the first 9 runs.  This option doesn't work.

*Option 2:* Return a complicated data structure (an arrayref of hashrefs) which contains the success/failure of the task on each server and the data returned from the task.  Or it could be an arrayref of objects.  This works but its a lousy solution.  Its difficult for users to discover simple things like whether the task ran successfully on all servers.  They would need to do some kind of complicated grep/map/loop.

*Option 3:* Event driven programming.  `do_task()` should use the same ui as `run()`.  If the task fails on any server, `do_task()` should die, but not until its tried to run the task against all the servers.
```
do_task $task_name, callback => sub {
    my $result = shift;
    $result->success;    # did the task die or not
    $result->task_name;
    $result->server_name;
    $result->exit_code;  # checks $? in case $task_name runs a shell command
};
```
Personally I think this ui makes the most sense for `do_task()`.  But it doesn't make sense for `run_task()` since `run_task()` can only run on a single server.   So I don't believe `do_task()` and `run_task()` should have the same ui.  Here is my idea for the ui for `run_task()`:
```
my $result = run_task $task_name, on => $server;
$result->success;       # did the task die or not
$result->task_name;
$result->server_name;
$result->exit_code;  # checks $? in case $task_name runs a shell command
```
Also `run_task()` should die if the task fails.

**Comment by ferki** [comment]
@kablamo: wow, thanks for the in-depth analysis! I think this would be even better to put into the wiki as it is more like a design discussion, than a direct answer to the original issue here. Basically (at least) those were the topics I referred to earlier on #697, when I mentioned "we're trying to solve too much different underlying problems on one issue".

1. `do_task` vs. `run_task`: I agree that if they _could be merged_, then they _should be merged_. I would place my vote on keeping `run_task` and deprecate `do_task`, but it's just my 2 cents.

2. I believe the future is to use Parallel::ForkManager only. This means, at one point we should make the transition. That could happen earliest in 2.0.0, until then we can't break backward compatibility, so we need to make sure _both_ forking methods are working. That being said, we can start migrating towards Parallel::ForkManager more, and make additional features depending on it to also depend on its feature flag (so it should be opt-in, plus the new features might speed up transition among users).

3. I also would vote for Option 2. Partly because of `exec_autodie`, partly because that's the current recommendation as well ("tasks should do error checking themselves and die if needed"). I'm not sure about the causation in the latter case though.
 So if a task doesn't die, it succeeded. I'm not sure how that affects `transaction`/`rollback` pairs, and/or parallel runs, and/or exacution via "task brokers" (like Gearman), but I guess we can ensure keeping everything in order by adding more tests if needed.

4. After agreeing to, and implementing no.1., the questions about the UI for `do_task` and `run_task` would be irrelevant (we said we would keep only one of them and deprecate the other). But in general, I'm thinking about basically the same: there should be "something", to where tasks can return their values. But maybe this should be optional and/or gatherable later (thinking about deferred execution, task queueing and/or retries upon failure, or pull style management, or "fire and forget" execution for long running tasks).

So the goal seems to be set. We also seem to have discovered all/most of the underlying complexity. So I'd like to encourage to break it down to smaller chunks, and solve them one-by-one.

Also, I love _discussions_, but again, I _fell like_ this issue is not a place for discussing all of the (unexpectedly) pulled in details.

**Comment by kablamo** [comment]
Cool.  I think the most fundamental issue which needs to get solved first is **How should we indicate if a task succeeded or failed**.  So thats what I'm working on now.  I believe my summary pr (https://github.com/RexOps/Rex/pull/697) will resolve this.  

In response to your feedback above for the summary pr:
1. It is missing an implementation for Parallel::ForkManager though.  I'll see if I can add that today.  
2. I don't believe the pr affects transactions.  But I should write up an analysis of that.
3. I don't believe the pr affects parallel execution at all.  But I should write up an analysis of that.
4. I don't know anything about "task brokers" so I will probably need you to help me out there.

I vote to move all further discussion about task success/failure to that pr: https://github.com/RexOps/Rex/pull/697

**Comment by ferki** [comment]
Erm, I'm a bit confused here - was it resolved fully by #697?
